### ISSUE TITLE
Use Dense layer with RaggedTensor

### ISSUE THREAD

**Issue by Azmisov** [keras-team/keras#14869]
Wondering what the recommended approach for dense layers is with RaggedTensor inputs.

I have tried a ragged=True Keras Input fed into a Keras Dense layer; I get an error though due to: `keras/layers/core.py, line 1245 ` using tensordot operation. If you are supposed to be able to use Dense layers w/ RaggedTensor, please let me know. _(I'd like to point out that theoretically, a tensordot implementation should be possible so long as the final dim is not ragged)_

My workaround for now is a custom layer that uses tf.multiply + tf.reduce_sum. Is that the best approach currently?

**Comment by haifeng-jin** [comment]
Hi @Azmisov, thanks for the issue.
Would you please paste your code for the custom layer?
For using ragged tensor, what shape would you expect the weights would be?

**Comment by Azmisov** [comment]
For kernel weight shape, `(outputs, inputs)`, with dims prepended for broadcasting. And RaggedTensor `(..., inputs)`. This is what I was doing:

```python
class Dense(L.Layer):
	""" Dense layer that works for RaggedTensor """
	def __init__(self, outputs:int=1, use_bias:bool=True, kernel_initializer=None, **kwargs):
		super().__init__(**kwargs)
		self.outputs = outputs
		self.use_bias = use_bias
		self.kernel_initializer = tf.keras.initializers.HeNormal() if kernel_initializer is None else kernel_initializer

	def build(self, input_shape):
		s = [1]*(len(input_shape)-1)
		kinit = self.kernel_initializer(s + [self.outputs, input_shape[-1]], tf.dtypes.float32)
		self.kernel = tf.Variable(kinit, trainable=True)
		if self.use_bias:
			binit = tf.zeros_initializer()(s + [self.outputs], tf.dtypes.float32)
			self.bias = tf.Variable(binit, trainable=True)

	def call(self, inp):
		out = tf.reduce_sum(tf.multiply(tf.expand_dims(inp, axis=-2), self.kernel), axis=-1)
		if self.use_bias:
			out += self.bias
		return out
```

**Comment by haifeng-jin** [comment]
To my understanding, you mean even the length of the shape of the input tensor is not fixed.
`Dense` layer is just designed to support 1-dimensional tensor input.
Your use case would be a new layer instead of a new feature to the `Dense` layer.

**Comment by Azmisov** [comment]
The rank of the input can be fixed, see the custom layer code I included before. And I'm not sure what you mean by `Dense` only supporting 1-d input, [as the documentation](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense) explicitly defines how the computation will work for arbitrary dimensions.

Here's an example of what I'm thinking:

```python
BATCHES = 4
CHANNELS = 8

# Non-ragged
data = tf.random.normal((BATCHES, 10, CHANNELS))
out = tf.keras.layers.Dense(16)(data)
print(f"{data.shape} -> {out.shape}")

# Ragged
data = tf.RaggedTensor.from_row_lengths(tf.random.normal((BATCHES*10, CHANNELS)), [12,8,7,13])
# simulated Dense layer...
kernel = tf.random.normal((1, 1, 16, CHANNELS))
out = tf.reduce_sum(tf.multiply(tf.expand_dims(data, -2), kernel), -1)
print(f"{data.shape} -> {out.shape}")
```
Output:
```
(4, 10, 8) -> (4, 10, 16)
(4, None, 8) -> (4, None, 16)
```
That's for rank 3 tensors, but it would work with any rank as long as the last dimension is not ragged.

**Comment by haifeng-jin** [comment]
I believe RaggedTensor can be used with Dense layer in tf-nightly.
You may `pip install tf-nightly` to have a try.

Thanks

**Comment by haifeng-jin** [comment]
The implementation is here:
https://github.com/keras-team/keras/blob/master/keras/layers/core.py#L1203-L1206
